services/swagger/swagger.service.js

```diff
@@ -1,67 +1,69 @@
 'use strict'
 
-const LegacyService = require('../legacy-service')
-const { makeBadgeData: getBadgeData } = require('../../lib/badge-data')
+const Joi = require('joi')
+const BaseJsonService = require('../base-json')
 
-// For a Swagger Validator.
-module.exports = class Swagger extends LegacyService {
-  static registerLegacyRouteHandler({ camp, cache }) {
-    camp.route(
-      /^\/swagger\/(valid)\/(2\.0)\/(https?)\/(.+)\.(svg|png|gif|jpg|json)$/,
-      cache((data, match, sendBadge, request) => {
-        // match[1] is not used                 // e.g. `valid` for validate
-        // match[2] is reserved for future use  // e.g. `2.0` for OpenAPI 2.0
-        const scheme = match[3] // e.g. `https`
-        const swaggerUrl = match[4] // e.g. `api.example.com/swagger.yaml`
-        const format = match[5]
+const validatorSchema = Joi.object()
+  .keys({
+    schemaValidationMessages: Joi.array().items(
+      Joi.object({
+        level: Joi.string().required(),
+        message: Joi.string().required(),
+      }).required()
+    ),
+  })
+  .required()
 
-        const badgeData = getBadgeData('swagger', data)
+module.exports = class SwaggerValidatorService extends BaseJsonService {
+  static render({ message, clr }) {
+    return { message: message, color: clr }
+  }
+
+  static get url() {
+    return {
+      base: 'swagger/valid/2.0',
+      format: '(http(?:s)?)/(.+)',
+      capture: ['protocol', 'url'],
+    }
+  }
+
+  static get defaultBadgeData() {
+    return { label: 'swagger' }
+  }
 
-        const urlParam = encodeURIComponent(scheme + '://' + swaggerUrl)
-        const url = 'http://online.swagger.io/validator/debug?url=' + urlParam
-        const options = {
-          method: 'GET',
-          url: url,
-          gzip: true,
-          json: true,
-        }
-        request(options, (err, res, json) => {
-          try {
-            if (
-              err != null ||
-              res.statusCode >= 500 ||
-              typeof json !== 'object'
-            ) {
-              badgeData.text[1] = 'inaccessible'
-              sendBadge(format, badgeData)
-              return
-            }
+  async handle({ protocol, url }) {
+    const json = await this.fetch({ protocol, urlF: url })
+    const valMessages = json.schemaValidationMessages
 
-            const messages = json.schemaValidationMessages
-            if (messages == null || messages.length === 0) {
-              badgeData.colorscheme = 'brightgreen'
-              badgeData.text[1] = 'valid'
-            } else {
-              badgeData.colorscheme = 'red'
+    if (!valMessages || valMessages.length === 0) {
+      return this.constructor.render({ message: 'valid', clr: 'brightgreen' })
+    } else {
+      return this.constructor.render({ message: 'invalid', clr: 'red' })
+    }
+  }
+
+  async fetch({ protocol, urlF }) {
+    const url = `http://online.swagger.io/validator/debug?url=${protocol}://${urlF}`
```

Comment:
Could we improve on the old implementation and use URL parameters here? Here's an [example](https://github.com/badges/shields/blob/99ec8fa239f517a0a6383c7d310ef145d953fad9/services/wercker/wercker.service.js#L31) (`qs` option).